cmd/opampsupervisor/supervisor/supervisor.go

```diff
@@ -384,24 +435,64 @@ func (s *Supervisor) startOpAMP() error {
 		},
 		Capabilities: s.config.Capabilities.SupportedCapabilities(),
 	}
-	err = s.opampClient.SetAgentDescription(s.agentDescription)
-	if err != nil {
+	ad := s.agentDescription.Load().(*protobufs.AgentDescription)
+	if err = s.opampClient.SetAgentDescription(ad); err != nil {
 		return err
 	}
 
-	err = s.opampClient.SetHealth(&protobufs.ComponentHealth{Healthy: false})
-	if err != nil {
+	if err = s.opampClient.SetHealth(&protobufs.ComponentHealth{Healthy: false}); err != nil {
 		return err
 	}
 
 	s.logger.Debug("Starting OpAMP client...")
+	if err = s.opampClient.Start(context.Background(), settings); err != nil {
+		return err
+	}
+	s.logger.Debug("OpAMP client started.")
+
+	return nil
+}
+
+// startOpAMPServer starts an OpAMP server that will communicate
+// with an OpAMP extension running inside a Collector to receive
+// data from inside the Collector. The internal server's lifetime is not
+// matched to the Collector's process, but may be restarted
+// depending on information received by the Supervisor from the remote
+// OpAMP server.
+func (s *Supervisor) startOpAMPServer() error {
+	s.opampServer = server.New(newLoggerFromZap(s.logger))
 
-	err = s.opampClient.Start(context.Background(), settings)
+	var err error
+	s.opampServerPort, err = s.findRandomPort()
 	if err != nil {
 		return err
 	}
 
-	s.logger.Debug("OpAMP Client started.")
+	s.logger.Debug("Starting OpAMP server...")
+
+	err = s.opampServer.Start(newServerSettings(flattenedSettings{
+		endpoint:         fmt.Sprintf("localhost:%d", s.opampServerPort),
+		onConnectingFunc: func(_ *http.Request) {},
+		onMessageFunc: func(_ serverTypes.Connection, message *protobufs.AgentToServer) {
+			s.logger.Debug("Received message")
+			if message.AgentDescription != nil {
+				s.setAgentDescription(message.AgentDescription)
+			}
+			if message.EffectiveConfig != nil {
+				s.logger.Debug("Setting confmap")
+				s.effectiveConfig.Store(string(message.EffectiveConfig.ConfigMap.ConfigMap[""].Body))
```

Comment:
We should probably do a check here that `ConfigMap` has the `""` key, right?

Reply:
Good call, I've added a check and an error log if the empty string key can't be found.